proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,262 @@
+<!-- {% raw %} -->
+
+# Strict Availability Diagnostics
+
+* Proposal: [INF-NNNN](INF-NNNN-availability-diagnostics.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Sponsor: [Chris Bieneman](https://github.com/llvm-beanz)
+* Status: **Under Consideration**
+* Planned Version: 20YY (Clang-only)
+
+## Introduction
+
+Strict availability diagnostics introduces two new modes for diagnosing the use
+of unavailable shader APIs. Unavailable shader APIs are APIs that are exposed in
+HLSL code but are not available in the target shader stage or shader model
+version.
+
+## Motivation
+
+Today the enforcement of API availability is the responsibility of bytecode
+validators, runtimes and drivers operating on optimized shader code. Depending
+contextually on where this enforcement occurs this can result in errors being
+produced either late in compilation when mapping back to source locations is
+difficult or at runtime when the runtime or driver rejects the compiled shader.
+
+Strict availability diagnostics provides more actionable diagnostics for users
+by both having detailed error messages and accurate message locations. It also
+provides diagnostics early and allows implementations to generate meaningful
+diagnostics even without generation of debug information.
+
+> Note: The DXIL validator requires line table information to generate error
+> locations which can be inaccurate. Disabling line table information can have a
+> significant compile-time benefit.
+
+## Proposed solution
+
+### Availability Annotations
+
+In Clang, most shader APIs have header declarations that live in the default
+included `hlsl_intrinsics.h`. All shader API declarations have Clang
+availability attributes denoting version information for when the APIs are
+introduced, deprecated, and removed. This version annotation can be per-stage or
+for all stages. The availability annotations will need to exist on declarations
+regardless of this proposal, as they provide information to AST-based tooling
+such as IDE workflows.
+
+As an example of the annotations in action take the following declaration from
+`hlsl_intrinsics.h`:
+
+```c++
+__attribute__((availability(shadermodel, introduced = 6.0)))

```

Comment:
Shall we have a similar decoration for Vulkan? (Or other APIs?)
```
__attribute__((availability(vulkan, introduced = 1.1)))
```

Reply:
Aren't there constraints on availability of these intrinsics in different execution environments for Vulkan as well, similar to different shader stages in DX?  Wouldn't we want to express that here as well?

Reply:
I did update the doc with a `vulkan` annotation.

Reply:
Yes, I see that.  I was thinking that if we had shadermodel_pixel, and there are similar (but slightly different) constraints for Vulkan, like for "fragment", then wouldn't we expect some breakdown of `vulkan` into `vulkan_fragment`, etc...?  Or am I misunderstanding how this sort of thing would be described?

Reply:
If there are APIs that are restricted by vulkan execution model (the SPIR-V terminology for DX shader stages), we should add appropriate vulkan values. I did a quick scan of the SPIR-V and GLSL extensions and nothing stood out, but it was by no means exhaustive.

Reply:
Yeah, this is fine for now.

I'm not sure what you meant by scanning extensions.  Aren't there plenty of operations only available under certain execution models (not needing to consider extensions)?  Like the equivalent of ddx is probably not available under all execution models, right?

Reply:
Thanks Tex! Yes, we'd also need to have a `_stage` suffix for the availability, as you have for `shadermodel_`.